### PR TITLE
rewards split

### PR DESCRIPTION
**Pull request by dzmitry-lahoda**
so we need to split users position. he can do that by transferring share with rewards

### CODE REVIEW

File: rewards/src/lib.rs [hunk 1]
```diff
@@ -10,6 +10,7 @@ use orml_traits::RewardHandler;
 use scale_info::TypeInfo;
 use sp_core::U256;
 use sp_runtime::{
+	helpers_128bit::multiply_by_rational,
```
Comment (zqhxuyuan) [inline rewards/src/lib.rs:13]: use `multiply_by_rational_with_rounding` instead of `multiply_by_rational`.

Comment (dzmitry-lahoda) [reply rewards/src/lib.rs:13]: hm, seems it needs to upgrade polkadot deps. is not it? should than this PR merged after update OR can live with multiply_by_rational OR can make it work with u256?

Comment (dzmitry-lahoda) [reply rewards/src/lib.rs]: not sure, used u256 directly

File: rewards/src/lib.rs [hunk 2]
```diff
@@ -315,6 +319,53 @@ impl<T: Config> Pallet<T> {
...
+	) -> DispatchResult {
+		SharesAndWithdrawnRewards::<T>::mutate(pool, other, |increased_share| {
+			let (increased_share, increased_rewards) = increased_share;
+			SharesAndWithdrawnRewards::<T>::mutate_exists(pool, who, |share| {
+				if let Some((share, rewards)) = share {
+					if move_share < *share {
+						for (reward_currency, balance) in rewards {
+							let move_balance = U256::from(balance.to_owned().saturated_into::<u128>())
+								* U256::from(move_share.to_owned().saturated_into::<u128>())
+								/ U256::from(share.to_owned().saturated_into::<u128>());
```
Comment (xlc) [inline rewards/src/lib.rs:343]: need some comment to explain why this is safe

Comment (dzmitry-lahoda) [reply rewards/src/lib.rs]: explained some, so i feel back about that Share is not capped to some number so and saturated used in many places (thins will get broken if share is > u128, not only in this method so). i feel like Share needs more traits. they are not related mbr or u256, but converte into u128 said first.

Comment (dzmitry-lahoda) [reply rewards/src/lib.rs]: so this method seems as safe as others, until replace all sat_into_u128 with try_into or extend share traits. will close comment for now if that is ok

File: rewards/src/lib.rs [hunk 3]
```diff
@@ -41,8 +41,17 @@ where
 
 pub use module::*;
 
+#[macro_export]
+macro_rules! ensure_some {
```
Comment (xlc) [inline rewards/src/lib.rs:45]: this is unnecessary

just `let x = x.ok_or(err)?;`

**Comment by dzmitry-lahoda** [comment]
cc @xlc @zqhxuyuan

**Review by xlc — APPROVED** [review]

**Pull request merged**